### PR TITLE
feat(quantic): add focus trap to preview modal

### PR DESCRIPTION
**Pull request by MSLI-Coveo**
[SFINT-4497](https://coveord.atlassian.net/browse/SFINT-4497)
add focus trap to preview modal.
Pressing Shift+Tab on close button now sets focus to last focusable element of the quick view footer or the quick view  content or to header if no focusable element exists in content and footer.

#### Example: Quick view with focusable footer
https://user-images.githubusercontent.com/86681870/191519679-92b19f17-15d2-4897-8332-cc1a900161e8.mov

#### Example: Quick view without focusable footer
https://user-images.githubusercontent.com/86681870/191520152-5d4fbafc-441a-4ad8-8d6d-dd178bd0146b.mov

#### Example: Quick view without focusable content or focusable footer
https://user-images.githubusercontent.com/86681870/191519976-aeaf3fad-712d-438c-a18e-314534b53379.mov

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js [hunk 1]
```diff
@@ -259,4 +259,38 @@ export default class QuanticResultQuickview extends LightningElement {
       focusTarget.focus();
     }
   }
+
+  getLsatFocusableElement(node) {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:263]:
```suggestion
  getLastFocusableElement(node) {
```

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js [hunk 2]
```diff
@@ -259,4 +259,38 @@ export default class QuanticResultQuickview extends LightningElement {
...
+    const lastFocusable = Array.from(node.childNodes)
+      .map((item) => this.getLsatFocusableElement(item))
+      .filter((item) => !!item);
+
+    if (lastFocusable.length > 0) {
+      return lastFocusable[lastFocusable.length - 1];
+    } else if (isFocusable(node)) {
+      return node;
+    }
+    return null;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:275]: Put this in a method exposed in Utils

```suggestion
    const quickviewContainer = this.template.querySelector('.quickview__content-container');
    if(quickviewContainer) {
        return getLastFocusableElement(quickviewContainer);
    }
```

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js (hunk 1 above)
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:263]:
```suggestion
  getLastFocusableElement(node) {
```

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 3]
```diff
@@ -513,4 +513,36 @@ export function AriaLiveRegion(regionName, elem, assertive = false) {
...
+  // Source: https://stackoverflow.com/a/30753870
+  if (element.getAttribute('tabindex') === '-1') {
+    return false;
+  }
+  if (element.hasAttribute('tabindex')) {
+    return true;
+  }
+  if (element.getAttribute('contentEditable') === 'true') {
+    return true;
+  }
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:533]:
```suggestion
  if (element.hasAttribute('tabindex') || element.getAttribute('contentEditable') === 'true') {
    return true;
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js [hunk 4]
```diff
@@ -259,4 +259,38 @@ export default class QuanticResultQuickview extends LightningElement {
...
   }
+
+  getLsatFocusableElement(node) {
+    if (!node || node.nodeType === 3) return null;
+
+    const lastFocusable = Array.from(node.childNodes)
+      .map((item) => this.getLsatFocusableElement(item))
+      .filter((item) => !!item);
+
+    if (lastFocusable.length > 0) {
```
Comment (nathanlb) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:270]:
```suggestion
    if (lastFocusable.length) {
```

**Review by nathanlb — APPROVED** [review]
Nice!

**Review by erocheleau — CHANGES_REQUESTED** [review]
Hey that looks simple enough! Pretty neat solution, just a few adjustments

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js (hunk 1 above)
Comment (MSLI-Coveo) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js]: typo fixed

Comment (MSLI-Coveo) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js]: typo fixed

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 5]
```diff
@@ -513,4 +513,54 @@ export function AriaLiveRegion(regionName, elem, assertive = false) {
...
+}
+
+/**
+ * Returns the last focusable element of for an HTML element. 
+ * This function would NOT work with shadow root.
+ * @param {Element | null} element
+ * @returns {Element | null}
+ */
+ export function getLastFocusableElement(element) {
+  if (!element || element.nodeType === 3) return null;
```
Comment (lbergeron) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:554]: Please use the constant instead. See https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#value

```suggestion
  if (!element || element.nodeType === Node.TEXT_NODE) return null;
```

**Review by lbergeron — APPROVED** [review]

**Review by aelfaleh — APPROVED** [review]

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js [hunk 6]
```diff
@@ -250,13 +251,49 @@ export default class QuanticResultQuickview extends LightningElement {
...
+      }
+    }
+  }
+
+  get lastFocusableElementInFooterSlot() {
+    /** @type {HTMLSlotElement} */
+    const footerSlot = this.template.querySelector('slot[name=footer]');
+    if (footerSlot) {
+        const lastElement = getLastFocusableElement(footerSlot);
+        if (lastElement) return lastElement;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js:286]: won't `lastElement` be null anyway if there is none?
Just `return lastElement;`

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 7]
```diff
@@ -562,8 +562,12 @@ export function getLastFocusableElement(element) {
     return null;
   }
 
+  if (element.tagName === "SLOT" && /** @type {HTMLSlotElement} */(element).assignedElements().length) {
+    return getLastFocusableElementFromSlot(/** @type {HTMLSlotElement} */(element))
+  }
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:567]: Not sure I like the inline typing comment :D Is there a way to not need that?

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 8]
```diff
@@ -562,8 +562,12 @@ export function getLastFocusableElement(element) {
...
+
   const focusableElements = Array.from(element.childNodes)
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:569]: Wouldn't it be possible to just mark the `focusableElements` as `@type {HTMLElement[]}` instead?

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 9]
```diff
@@ -590,8 +594,12 @@ export function getFirstFocusableElement(element) {
     return null;
   }
 
+  if (element.tagName === "SLOT" && /** @type {HTMLSlotElement} */(element).assignedElements().length) {
+    return getFirstFocusableElementFromSlot(/** @type {HTMLSlotElement} */(element))
+  }
+
   const focusableElements = Array.from(element.childNodes)
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:601]: Same as above

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 10]
```diff
@@ -613,3 +621,35 @@ function isCustomElement(element) {
   }
   return false;
 }
+
+/**
+ * Returns the last focusable element in an HTML slot.
+ * @param {HTMLSlotElement | null} slotElement
+ */
+function getLastFocusableElementFromSlot(slotElement) {
+  if (!slotElement && slotElement.assignedElements) return null;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:630]:
```suggestion
  if (!slotElement && slotElement.assignedElements) {
    return null;
  }
```

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 11]
```diff
@@ -613,3 +621,35 @@ function isCustomElement(element) {
...
+  const focusableElements = Array.from(slotElement.assignedElements())
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:631]: Same as above: `type HTMLElement[]` instead?

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js [hunk 12]
```diff
@@ -613,3 +621,35 @@ function isCustomElement(element) {
...
+  }
+  return null;
+}
+
+/**
+ * Returns the first focusable element in an HTML slot.
+ * @param {HTMLSlotElement | null} slotElement
+ */
+function getFirstFocusableElementFromSlot(slotElement) {
+  if (!slotElement && slotElement.assignedElements) return null;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:646]:
```suggestion
  if (!slotElement && slotElement.assignedElements) {
    return null;
  }
```

**Review by erocheleau — CHANGES_REQUESTED** [review]

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js (hunk 8 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js:569]: Actually the problem here is not the type of the array `focusableElements`. the problem is in the return value of `element.childNodes` which returns an array of type `ChildNodes[]`.
The values of `element.childNodes` are then passed to the function `getLastFocusableElement` which only accepts the type `HTMLElement`. 
When I try to mark `element.childNodes` as `@type {HTMLElement[]}`, I get the following error: `Type 'ChildNode[]' is not assignable to type 'HTMLElement[]'`
That's why I'm using this inline type annotation, its the only way around that I found.

File: packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js (hunk 7 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticUtils/quanticUtils.js]: inline typing removed 👌

**Review by erocheleau — APPROVED** [review]

**Pull request merged**